src/metabase/upload.clj

```diff
@@ -179,75 +165,62 @@
   (boolean (re-matches #"(?i)true|t|yes|y|1|false|f|no|n|0" s)))
 
 (defn- boolean-or-int-string? [s]
-  (boolean (#{"0" "1"} s)))
+  (contains? #{"0" "1"} s))
+
+(defn- varchar-255? [s]
+  (<= (count s) 255))
 
 ;; end [[value->type]] helpers
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
+(def ^:private non-inferable-types
+  #{::auto-incrementing-int-pk})
+
+(defn- assert-all-types-checks-present
+  [type->check]
+  (assert (every? type->check (remove non-inferable-types value-types))
+          "Every value-type must be registered in settings->type->check")
```

Comment:
The dev ex felt a bit better to me - instant feedback. This is a classic invariant that you'd catch with a static type checker.

Perhaps not a factor here, but when a bunch of distant tests break due to some trivial invariant being violated it's waste of time to read through the failures to diagnose. I guess a malli schema could solve this, but I still think a "compile time" error is a better experience.

Can shunt to a test if it feels deeply wrong to you - I'm still finding my style in this ecosystem.

Reply:
I guess this feels a little strange because it's proving some property of `settings->type->check`, not checking the pre/post-conditions of each function call. 

Run-time assertions can be costly, but in this case it's not that bad because you're only running this assertion once per upload, which is a slow operation anyway.

I'm inclined to discourage this style in general, but at the same time this specific case is fine and not worth bikeshedding over

Reply:
I agree that it feels strange - the shape of the map is not even dynamic. It's pity that it exists only to close over the settings. I thought about just having a static map whose functions all took settings as well, but then we'd be rebuilding the regular expressions constantly.

I've been assuming that assertions are disabled in production, I've never worked on a production Java app that leaves them enabled, but I guess we don't control how people run the JAR.

I'll send a small PR with using Malli instead. It's more declarative and always off in production. The error message seems unhelpful, but I'll chat to a Malli guru about that.

Reply:
PR: [Use malli to ensure all column types are covered](https://github.com/metabase/metabase/pull/39612)